Modern fragments: text content elements (part 2)

m-vo opened the pull request:
Followup to #4775 

This PR adds the missing text content elements:
* `content_element/text`
* `content_element/table`

It also improves the usage of the `HtmlAttributes` class by allowing conditions everywhere.

Todo:

- [x] add tests
- [x] adjust/inline table sorting script

~~**Note:** This PR is currently based on #4775 and #4730; needs to be rebased once merged.~~

m-vo committed:
squashed commits from #4775 + #4730

m-vo committed:
normalize name of context variable (should be snake case)

m-vo committed:
add image, gallery and list content elements

m-vo committed:
add link content elements

m-vo committed:
allow arbitrary conditions everywhere in the HtmlAttributes class

m-vo committed:
add text + table content element

ausi commented on core-bundle/src/String/HtmlAttributes.php:
```diff
@@ -145,11 +147,20 @@ public function unset(string $name, \Stringable|bool|int|string|null $condition
+     *
+     * If a falsy $condition is specified, the method is a no-op.
+     */
+    public function addClass(string $classes, mixed $condition = true): self
```
Just a thought: how about allowing an array of strings too?

m-vo replied:
Ah, you mean `string|array $classes`? I like :+1:

m-vo replied:
see fa0b4690c3778a48e23d1498f0897959940d189b

ausi commented on core-bundle/src/Controller/ContentElement/TableController.php:
```diff
@@ -0,0 +1,51 @@
+        // Table fields
+        $rows = StringUtil::deserialize($model->tableitems, true);
+        $header = $model->thead && !empty($rows) ? array_shift($rows) : [];
+        $footer = $model->tfoot && !empty($rows) ? array_pop($rows) : [];
```
```suggestion
        $header = $model->thead && $rows ? array_shift($rows) : [];
        $footer = $model->tfoot && $rows ? array_pop($rows) : [];
```
Nitpick ☺️

ausi commented on core-bundle/src/Controller/ContentElement/TableController.php:
```diff
@@ -0,0 +1,51 @@
+        $template->set('caption', $model->summary ?: null);
+
+        // Client side sorting
+        $sorting = empty($header) || !$model->sortable ? null : [
```
```suggestion
        $sorting = !$header || !$model->sortable ? null : [
```

ausi commented on core-bundle/src/Resources/contao/templates/_new/content_element/table.html.twig:
```diff
@@ -0,0 +1,79 @@
+        .set('id', 'table_' ~ data.id, sorting)
+        .addClass('sortable', sorting)
+        .set('data-sort-default', sorting.column|default ~ '|' ~ sorting.order|default, sorting)
+    %}
```
```suggestion
    {% set table_attributes = attrs(table_attributes|default) %}
    {% if sorting %}
        {% table_attributes
            .set('id', 'table_' ~ data.id)
            .addClass('sortable')
            .set('data-sort-default', sorting.column|default ~ '|' ~ sorting.order|default)
        %}
    {% endif %}
```
Would something like this be more readable?

m-vo replied:
Had this before, but then changed it to the inline version. :smile: 

But already convinced. I used the `{% do %}` tag in 69828a6251c752e1d073c458ec3944c7d0fbb7b2.

ausi commented on core-bundle/src/Resources/contao/templates/_new/content_element/table.html.twig:
```diff
@@ -0,0 +1,79 @@
+                <thead>
+                {% block header_inner %}
+                    <tr>
+                        {% for col in header %}
```
Maybe rename `col` to `cell` everywhere in this template?

m-vo replied:
see ea7409be00c212879e48208b6dd86f63f8b89ee2

m-vo committed:
Update core-bundle/src/Controller/ContentElement/HyperlinkController.php

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

m-vo committed:
Update core-bundle/src/Controller/ContentElement/TableController.php

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

m-vo committed:
Update core-bundle/src/Controller/ContentElement/TableController.php

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

m-vo committed:
use a block instead of inline configuration

m-vo committed:
rename col to cell

m-vo committed:
allow arrays alongside strings when adding/removing classes

ausi reviewed (dismissed):
Approved, but #4775 and #4730 need to be merged first.

m-vo committed:
simplify toplink script + make it forward compatible (thanks @ausi)

m-vo committed:
add and improve tests + small adjustments

leofeyer commented on core-bundle/src/Controller/ContentElement/TableController.php:
```diff
@@ -0,0 +1,51 @@
+        $template->set('caption', $model->summary ?: null);
+
+        // Client side sorting
+        $sorting = !$header || !$model->sortable ? null : [
```
```suggestion
        $sorting = (!$header || !$model->sortable) ? null : [
```

Not sure if the operator precedence is correct otherwise.

m-vo replied:
`||` has a higher precedence than `? :`, so it should work, But I agree, that the parantheses make it more clear, so let's go with it. :+1:

m-vo replied:
Or maybe not. The CS fixer (`UselessParenthesesSniff`) removed it again :smile:

m-vo committed:
Merge remote-tracking branch 'origin/5.x' into twig/content-elements/links

m-vo committed:
CS

m-vo committed:
fix merge errors + make forward compatible

m-vo committed:
clarify expression

Co-authored-by: Leo Feyer <github@contao.org>

m-vo committed:
Merge branch 'twig/content-elements/links' into twig/content-elements/texts-2

m-vo committed:
CS

m-vo committed:
Merge remote-tracking branch 'origin/5.x' into twig/content-elements/texts-2

m-vo commented:
This is rebased and IMHO ready to go now as well. :+1:

m-vo committed:
make table sortable with vanilla js + extract proper component

m-vo committed:
homogenize script usage across components

m-vo commented:
I updated the PR to include an accessible vanilla table sorting script (https://www.dhs.state.il.us/page.aspx?item=55196).

I also refactored the table creation into a neat component. Now you could even create tables on the fly:
```twig
{% with {rows: [['foo', 'bar'], ['baz','foobar']], sorting: true} %}
  {{ block('table_component') }}
{% endwith %}
```

/cc @ausi

ausi commented on core-bundle/src/Resources/contao/templates/_new/component/_table.html.twig:
```diff
@@ -0,0 +1,196 @@
+                        const instances = new WeakMap();
+                        const initTree = (root) => {
+                            const elements = root.tagName === 'TABLE' && root.hasAttribute('data-sortable-table') ?
+                                [root] : root.querySelectorAll('table[data-sortable-table]');
```
```suggestion
                            const elements = root.matches('table[data-sortable-table]') ?
                                [root] : root.querySelectorAll('table[data-sortable-table]');
```

m-vo committed:
add a test for the TableController

ausi commented on core-bundle/src/Resources/contao/templates/_new/component/_table.html.twig:
```diff
@@ -0,0 +1,196 @@
+
+                        const script = document.createElement('script');
+                        script.setAttribute('src', 'bundles/contaocore/sortabletable.js');
+                        script.setAttribute('async', true);
```
```suggestion
                        script.src = 'bundles/contaocore/sortabletable.js';
                        script.async = true;
```

ausi commented on core-bundle/src/Resources/contao/templates/_new/component/_table.html.twig:
```diff
@@ -0,0 +1,196 @@
+                                attributeFilter: ['data-sortable-table'],
+                                childList: true,
+                                subtree: true
+                            });
```
```suggestion
                            }).observe(document.documentElement, {
                                childList: true,
                                subtree: true
                            });
```

Observing attributes is overkill here I think and not worth the performance cost IMO.

ausi commented on core-bundle/src/Resources/contao/templates/_new/component/_table.html.twig:
```diff
@@ -0,0 +1,196 @@
+                        })
+
+                        document.head.appendChild(script);
+                        document.addEventListener('domready', () => initTree(document));
```
I think we should remove the `domready` event here and call `initTree(document)` in the script onload listener instead.

m-vo replied:
But then the script might be loaded before the DOM is ready and we might miss the initial elements. Or can't this be the case?

m-vo replied:
Convinced. :relaxed: 
See ad12295de57854252bdf9817a0f47cdc6dc03690.

ausi commented on core-bundle/src/Resources/contao/templates/_new/component/_table.html.twig:
```diff
@@ -0,0 +1,196 @@
+{#
+    This component outputs a table with client side sorting capabilities.
+
+        <table data-sortable-table="1|asc">
```
This format seems odd, should we use a space seperated string, which is more common in HTML attributes?

```suggestion
        <table data-sortable-table="1 asc">
```

Or two attributes?

```suggestion
        <table data-sortable-table="1" data-sortable-table-direction="asc">
```

m-vo replied:
Agreed, I just used this because we had this in the old element. Changed in c6e5938283edc05dcbbd9adf8aa74aca04ecfbbf

m-vo committed:
polish the JS

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

m-vo committed:
use space as separation character

m-vo committed:
CS

m-vo committed:
simplify initialization

leofeyer reviewed (changes requested):
The changes are perfect, however, we cannot ship the `sortable.js` script this way. On the one hand, it should be a reusable component, and on the other hand, it should also be used e.g. here:

https://github.com/contao/contao/blob/47697073752ae4d6fa6b63b28e0b54c50a8143f3/core-bundle/src/Resources/public/modulewizard.js#L23-L31

leofeyer committed:
CS

m-vo committed:
only show the "includeTemplates" hint for the legacy content element

m-vo committed:
CS

ausi reviewed (dismissed):
Only open question is IMO which script to use for `sortabletable.js` and how to distribute that script (e.g. via contao-components).

I’d be happy with the what’s currently in sortabletable.js too.

m-vo committed:
use tristen/tablesort for sorting

m-vo commented:
I switched to [tristen/tablesort](https://github.com/tristen/tablesort) in b06d17381f519f5989493bb50c71db7434fb4342. 

For the purpose of testing, the script is still in the resources folder - we would need to remove it from there and adjust the path to a contao-component once released as one.

ausi reviewed (dismissed):
Approved apart from the script needing to be moved to contao-components

m-vo committed:
get tristen/tablesort script from contao-components

m-vo committed:
Merge remote-tracking branch 'origin/5.x' into twig/content-elements/texts-2

m-vo commented:
With fffb3e943c1a85ecc7e4d4f79c756bd7384ba701 we're now using https://github.com/contao-components/tristen-tablesort instead of manually distributing the files.

m-vo committed:
Merge branch '5.x' into twig/content-elements/texts-2

leofeyer committed:
Merge branch '5.x' into twig/content-elements/texts-2

leofeyer commented:
Thank you @m-vo.
